### PR TITLE
v3.0.0: Site roles, program authz, user authz

### PR DESCRIPTION
**Pull request by daisieh**
## v3.0.0: Site roles, program authz, user authz
* Site roles, including site admin, now defined in Opa
* Program authorizations are defined in Opa's vault secret store
* User-specific program authorizations are defined in Opa's vault secret store
* Refactored rego policies and created unit tests

- [x] Tagged as a release in this repo
- [x] Passes integration tests on a development instance
- [x] Images pushed to Dockerhub

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -28,6 +28,10 @@ USER candig
...
 
+RUN curl -L -o opa https://openpolicyagent.org/downloads/v0.63.0/opa_linux_amd64_static
+
+RUN chmod 755 ./opa
+
 RUN touch /app/initial_setup
 
-ENTRYPOINT ["bash", "/app/entrypoint.sh"]
\ No newline at end of file
+ENTRYPOINT pytest
```
Comment (kcranston) [inline Dockerfile:37]: can you explain this change? I don't understand the subsitution of entrypoint.sh with pytest

File: Dockerfile [hunk 2]
```diff
@@ -28,6 +28,10 @@ USER candig
 
 WORKDIR /app/
 
+RUN curl -L -o opa https://openpolicyagent.org/downloads/v0.63.0/opa_linux_amd64_static
```
Comment (kcranston) [inline Dockerfile:31]: how does this intersect with the `image` in https://github.com/CanDIG/CanDIGv2/blob/develop/lib/opa/docker-compose.yml ? Would mac and windows users have to change this url manually?

**Comment by kcranston** [comment]
Thanks for this! Couple of questions:
* for data ingest workflow, how does the Program data get created in opa? Would the user uploading a new Program have a site `curator` role which allows for adding new Program data, or does ingest do this on the user's behalf using a service token? Or does a system admin need to create the program before a curator can upload data?
* is there a sister PR with new integration tests for these changes? or a ticket with the integration tests that should be written?

**Comment by daisieh** [comment]
Documentation for the ingest workflow was added in https://github.com/CanDIG/candigv2-ingest/pull/93. When ingesting a Program Authorization for the first time, Opa needs the user to be someone with permission to ingest any program for the site. At the moment, the only site role for that is Admin, but in future, we could add a site-curator role that would be allowed to do this as well. 

If the Program Authorization has not been ingested before data is ingested, the only user that would be allowed to ingest data for that program is the site admin. If the site admin adds the Program Authorization first, any listed program curator in the Program Authorization would then have the authz to ingest data.

Integration tests were added in https://github.com/CanDIG/CanDIGv2/pull/550, which was the sister PR to https://github.com/CanDIG/candig-opa/pull/52 and https://github.com/CanDIG/candigv2-ingest/pull/80.

File: Dockerfile (hunk 2 above)
Comment (daisieh) [reply Dockerfile:31]: It doesn't intersect at all. This one just lives in the opa-runner container and allows it to have an executable opa on its own, for pytest.

File: Dockerfile (hunk 1 above)
Comment (daisieh) [reply Dockerfile:37]: This is, in our stack context, always overwritten by the entrypoint command in the CanDIGv2/lib/opa/docker-compose.yml. That is the one that says 
```
    entrypoint:
      - "bash"
      - "/app/entrypoint.sh"
```

But in the case where you are building the Dockerfile directly, as in the Github action, we can execute just the pytest files and not set up an interactive opa server.

**Review by kcranston — APPROVED** [review]

**Pull request merged**